clippy_lints/src/manual_div_ceil.rs

```diff
@@ -0,0 +1,156 @@
+    /// ### What it does
+    /// Checks for an expression like `(x + (y - 1)) / y` which is a common manual reimplementation
+    /// of `x.div_ceil(y)`.
+    ///
+    /// ### Why is this bad?
+    /// It's simpler and more readable.
+    ///
+    /// ### Example
+    /// ```no_run
+    /// let x: i32 = 7;
+    /// let y: i32 = 4;
+    /// let div = (x + (y - 1)) / y;
+    /// ```
+    /// Use instead:
+    /// ```no_run
+    /// #![feature(int_roundings)]
+    /// let x: i32 = 7;
+    /// let y: i32 = 4;
+    /// let div = x.div_ceil(y);
+    /// ```
+    #[clippy::version = "1.81.0"]
+    pub MANUAL_DIV_CEIL,
+    complexity,
+    "manually reimplementing `div_ceil`"
+}
+
+pub struct ManualDivCeil {
+    msrv: Msrv,
+}
+
+impl ManualDivCeil {
+    #[must_use]
+    pub fn new(msrv: Msrv) -> Self {
+        Self { msrv }
+    }
+}
+
+impl_lint_pass!(ManualDivCeil => [MANUAL_DIV_CEIL]);
+
+impl<'tcx> LateLintPass<'tcx> for ManualDivCeil {
+    fn check_expr(&mut self, cx: &LateContext<'tcx>, expr: &Expr<'_>) {
+        if !self.msrv.meets(msrvs::MANUAL_DIV_CEIL) {
+            return;
+        }
+
+        let mut applicability = Applicability::MachineApplicable;
+
+        if let ExprKind::Binary(div_op, div_lhs, div_rhs) = expr.kind
+            && check_int_ty_and_feature(cx, div_lhs)
+            && check_int_ty_and_feature(cx, div_rhs)
+            && div_op.node == BinOpKind::Div
+        {
+            // (x + (y - 1)) / y
+            if let ExprKind::Binary(add_op, add_lhs, add_rhs) = div_lhs.kind
+                && add_op.node == BinOpKind::Add
+                && let ExprKind::Binary(sub_op, sub_lhs, sub_rhs) = add_rhs.kind
+                && sub_op.node == BinOpKind::Sub
+                && check_literal(sub_rhs)
+                && check_eq_expr(cx, sub_lhs, div_rhs)
+            {
+                build_suggestion(cx, expr, add_lhs, div_rhs, &mut applicability);
+                return;
+            }
+
+            // ((y - 1) + x) / y
+            if let ExprKind::Binary(add_op, add_lhs, add_rhs) = div_lhs.kind
+                && add_op.node == BinOpKind::Add
+                && let ExprKind::Binary(sub_op, sub_lhs, sub_rhs) = add_lhs.kind
+                && sub_op.node == BinOpKind::Sub
+                && check_literal(sub_rhs)
+                && check_eq_expr(cx, sub_lhs, div_rhs)
+            {
+                build_suggestion(cx, expr, add_rhs, div_rhs, &mut applicability);
+                return;
+            }
+
+            // (x + y - 1) / y
+            if let ExprKind::Binary(sub_op, sub_lhs, sub_rhs) = div_lhs.kind
+                && sub_op.node == BinOpKind::Sub
+                && let ExprKind::Binary(add_op, add_lhs, add_rhs) = sub_lhs.kind
+                && add_op.node == BinOpKind::Add
+                && check_literal(sub_rhs)
+                && check_eq_expr(cx, add_rhs, div_rhs)
+            {
+                build_suggestion(cx, expr, add_lhs, div_rhs, &mut applicability);
+            }
```

Comment:
This three blocks are almost identical, you could abstract them into an inline function

Reply:
Yes, was thinking about that, but I don't really see a way of abstracting this while still keeping it readable and maintainable, so I decided to keep it as is

Reply:
I've tinkered with this a bit, and it seems to be that kind of code similar enough to assume it does very similar things, while being just different that making it generic is impossible. (We could use a macro, but that would defeat the purpose of compacting it)

Reply:
So maybe we should keep it in the current way?

Reply:
Yeah, we should keep it the current way.